keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resolvers/ImageResolver.kt

```diff
@@ -79,44 +80,66 @@ class ImageResolver(
       deliveryConfig,
       artifact,
       environment.name
-    ) ?: throw NoImageSatisfiesConstraints(artifact.name, environment.name)
+    ) ?: throw NoArtifactVersionHasBeenApproved(artifact.name, environment.name)
 
     val images = imageService.getLatestNamedImages(
       appVersion = artifactVersion.parseAppVersion(),
       account = account,
       regions = regions
     )
 
-    return VersionedNamedImage(images, artifact, artifactVersion)
+    val imageIdByRegion: Map<String, String> = images
+      .mapValues { (region, image) -> image.amis[region]?.first() }
+      .filterNotNullValues()
+
+     val vmImages = images.mapValues { (region, namedImage) ->
+      VirtualMachineImage(
+        id = imageIdByRegion.getValue(region),
+        appVersion = namedImage.appVersion,
+        baseImageVersion = namedImage.baseImageVersion
+      )
+    }
+
+    val bakedVmImages = mutableMapOf<String, VirtualMachineImage>()
+    // we might have pre-knowledge of baked images here, so let's load that.
+    bakedImageRepository
+      .getLatestByArtfiactVerstion(artifactVersion, artifact)
+      ?.let { bakedImage ->
+        bakedImage.amiIdsByRegion.forEach{ (region, amiId) ->
+          bakedVmImages[region] = VirtualMachineImage(
+            id = amiId,
+            appVersion = bakedImage.appVersion,
+            baseImageVersion = bakedImage.baseAmiVersion
+          )
+        }
+      }
```

Comment:
You could simplify this a bit:
```suggestion
    // we might have pre-knowledge of baked images here, so let's load that.
    val bakedVmImages = bakedImageRepository
      .getLatestByArtfiactVerstion(artifactVersion, artifact)
      ?.let { bakedImage ->
        bakedImage.amiIdsByRegion.map{ (region, amiId) ->
          VirtualMachineImage(
            id = amiId,
            appVersion = bakedImage.appVersion,
            baseImageVersion = bakedImage.baseAmiVersion
          )
      }
      ?: emptyList()
```

Reply:
I was totally trying to get this to work, but I couldn't. then I forgot about it. the snippet you created is a list, but I need a map. This prompted me to figure it out using `bakedImage.amiIdsByRegion.mapValues {`. Thanks for the prompt!